### PR TITLE
[Data] Require `batch_size` for GPU `map_batches` in strict mode

### PR DESCRIPTION
**Pull request by amogkam**
In strict mode, require a batch_size for `map_batches` when requesting GPUs. This makes batching more explicit to users.

Reduce the CPU default batch size in strict mode from 4096 to 1024.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/datastream.py [hunk 1]
```diff
@@ -597,6 +597,11 @@ def map_batches(
 
         target_block_size = None
         if batch_size == "default":
+            ctx = DataContext.get_current()
+            if ctx.strict_mode:
+                raise ValueError(
+                    "`batch_size` must be specified when using strict " "mode "
+                )
```
Comment (c21) [inline python/ray/data/datastream.py:604]: I just figured there're seveal APIs depending on default batch_size, e.g. [random_sample](https://github.com/ray-project/ray/blob/master/python/ray/data/datastream.py#L1137), add_column, drop_columns, select_columns etc. We need to handle them properly.

File: python/ray/data/tests/test_strict_mode.py [hunk 2]
```diff
@@ -111,6 +111,12 @@ def f(x):
     assert isinstance(batch["id"], np.ndarray), batch
 
 
+def test_strict_require_batch_size(ray_start_regular_shared):
```
Comment (c21) [inline python/ray/data/tests/test_strict_mode.py:114]: let's also add unit test for all other APIs calling map_batches under the hood, e.g. https://github.com/ray-project/ray/pull/34588/files#r1171781083 .

File: python/ray/data/datastream.py [hunk 3]
```diff
@@ -597,6 +597,11 @@ def map_batches(
 
         target_block_size = None
         if batch_size == "default":
+            ctx = DataContext.get_current()
+            if ctx.strict_mode:
+                raise ValueError(
+                    "`batch_size` must be specified when using strict " "mode "
```
Comment (ericl) [inline python/ray/data/datastream.py:603]:
```suggestion
+                raise StrictModeError(
+                    "In strict mode, batch_size must be specified as an argument for "
+                    "`map_batches`, e.g., `map_batches(..., batch_size=256)`.")
```

File: python/ray/data/tests/test_strict_mode.py [hunk 4]
```diff
@@ -111,6 +111,12 @@ def f(x):
...
+    ds = ray.data.range(1)
+    with pytest.raises(ValueError):
```
Comment (ericl) [inline python/ray/data/tests/test_strict_mode.py:116]:
```suggestion
    with pytest.raises(StrictModeError):
```

**Comment by ericl** [comment]
Closing per offline discussion with @pcmoritz in favor of a warning for now...

File: python/ray/data/datastream.py (hunk 1 above)
Comment (amogkam) [reply python/ray/data/datastream.py]: Updated to still use a default for CPU `map_batches`

File: python/ray/data/context.py [hunk 5]
```diff
@@ -120,6 +120,9 @@
 # Default batch size for batch transformations.
 DEFAULT_BATCH_SIZE = 4096
 
+# Default batch size for batch transformations in strict mode.
+STRICT_MODE_DEFAULT_BATCH_SIZE = 1000
```
Comment (ericl) [inline python/ray/data/context.py:124]: 1024?

**Review by ericl — APPROVED** [review]

File: python/ray/data/tests/test_strict_mode.py (hunk 2 above)
Comment (amogkam) [reply python/ray/data/tests/test_strict_mode.py]: batch_size is no longer required for CPU `map_batches`

File: python/ray/data/tests/test_strict_mode.py (hunk 4 above)
Comment (amogkam) [reply python/ray/data/tests/test_strict_mode.py]: updated

File: python/ray/data/context.py (hunk 5 above)
Comment (amogkam) [reply python/ray/data/context.py]: sure, updated

File: python/ray/data/block.py [hunk 6]
```diff
@@ -174,6 +175,27 @@ def _apply_strict_mode_batch_format(given_batch_format: Optional[str]) -> str:
...
 
 
+def _apply_strict_mode_batch_size(
+    given_batch_size: Optional[Union[int, Literal["default"]]], use_gpu: bool
+) -> Optional[int]:
+    ctx = ray.data.DatasetContext.get_current()
+    if ctx.strict_mode:
+        if use_gpu and not given_batch_size or given_batch_size == "default":
+            raise StrictModeError(
+                "`batch_size` must be provided to `map_batches` when requesting GPUs."
```
Comment (pcmoritz) [inline python/ray/data/block.py:185]: Let's add a little more explanation here, something like
```
The optimal batch size for ML models depends on the model, data and GPU used -- try the setting that comes recommended with your model or example script, or try out which setting is the most performant for your situation and fits on GPU memory. Typically smaller batch sizes like 1, 2, 4, 8, 16, 32 are optimal for large models or memory constrained environments and larger batch sizes like 64, 128, 256 or 512 are optimal for smaller models or large GPUs.
```

**Review by pcmoritz — APPROVED** [review]
Thanks, I added a note to improve the error message, other than that it looks great!

File: python/ray/data/block.py (hunk 6 above)
Comment (pcmoritz) [reply python/ray/data/block.py:185]: We could also link a doc page that goes a little more into this and e.g. shows people how to use the dashboard to see GPU memory utilization (possibly even link to a GPU memory profiler or integrate one into the dashboard if available).

Comment (amogkam) [reply python/ray/data/block.py]: updated error message

**Comment by ericl** [comment]
Somehow this is changing the batch size even when strict mode is off in tests.

**Comment by amogkam** [comment]
Failing tests are also failing on master. Going to merge.

**Pull request merged**